service/history/execution/context.go

```diff
@@ -265,11 +269,28 @@ func NewContext(
 }
 
 func (c *contextImpl) Lock(ctx context.Context) error {
-	return c.mutex.Lock(ctx)
+	err := c.mutex.Lock(ctx)
+	if err != nil {
+		return err
+	}
+	c.lockTime = time.Now()
+	return nil
 }
 
 func (c *contextImpl) Unlock() {
+	elapsed := time.Since(c.lockTime)
 	c.mutex.Unlock()
+
+	if c.lockTime.IsZero() { // skip logging if the lock is never acquired
+		return
+	}
+
+	c.lockTime = time.Time{}
+	c.metricsClient.RecordTimer(metrics.WorkflowContextScope, metrics.WorkflowContextLockLatency, elapsed)
+	if elapsed > c.maxLockDuration {
+		c.maxLockDuration = elapsed
+		c.logger.Info("workflow context lock is released. this is logged only when it's longer than maxLockDuration", tag.WorkflowContextLockLatency(elapsed))
+	}
```

Comment:
Same as Lock(), it's safer to put logging logic before unlock.

Reply:
switched to atomics